crates/hir-ty/src/builder.rs

```diff
@@ -232,6 +232,28 @@ impl TyBuilder<()> {
         TyBuilder::new((), params, parent_subst)
     }
 
+    pub fn subst_for_closure(
+        db: &dyn HirDatabase,
+        parent: DefWithBodyId,
+        sig_ty: Ty,
+    ) -> Substitution {
+        Substitution::from_iter(
+            Interner,
+            parent
+                .as_generic_def_id()
+                .map(|p| {
+                    generics(db.upcast(), p)
+                        .placeholder_subst(db)
+                        .iter(Interner)
+                        .chain(iter::once(&GenericArgData::Ty(sig_ty).intern(Interner)))
```

Comment:
I know the ordering wouldn't matter in this case, but I'd like sig_ty to come before parent args for the sake of consistency with other substs.

Also, `sig_ty.cast(Interner)` is nicer compared to constructing `GenericArg` from `GenericArgData` directly imo.

Reply:
I want to eventually make closure subst equal to rustc (I wanted to not encode closure type and fields in the subst, but it looks like it is unavoidable long term), and rustc uses this ordering if I understand correctly. Is it consistent in rustc? If so, we probably need to change other substs to match this. If rustc is inconsistent (and has no good reason for it) I'm not opposed to diverge from rustc here.

Reply:
The thing is, chalk expects an item's own parameters/arguments before its parent's for *some* `Binder`s/`Substitution`s, which is different from rustc. I felt (and still do) it was more natural to order everything as such rather than special-casing those chalk puts constraints on, which led to #13335 and #13505.

Reply:
It make sense. Changed.